File: connectors/unified-search/app/sources/vikunja.py
"""Vikunja `tasks/all?s=` connector.

The user's token is relayed as-is (study 2.2 line 391): only tasks from
projects the user has access to are returned, with no permission logic
duplicated here.
"""

from __future__ import annotations

import os
from urllib.parse import quote

import httpx

from app.types import SearchResultItem

VIKUNJA_BASE_URL = os.environ.get("VIKUNJA_BASE_URL", "https://vikunja.example.org")


async def search_vikunja(
    query: str,
    user_token: str,
    client: httpx.AsyncClient | None = None,
) -> list[SearchResultItem]:
    owns_client = client is None
    client = client or httpx.AsyncClient()
    try:
        response = await client.get(
            f"{VIKUNJA_BASE_URL}/api/v1/tasks/all?s={quote(query)}",
            headers={"Authorization": f"Bearer {user_token}"},
        )
    finally:
        if owns_client:
            await client.aclose()

    if response.status_code >= 400:
        raise RuntimeError(f"Vikunja search failed with status {response.status_code}")

    data = response.json() or []

    return [
        SearchResultItem(
            source="vikunja",
            id=str(task.get("id", "")),
            title=task.get("title") or "(task)",
            snippet=task.get("description"),
            url=f"{VIKUNJA_BASE_URL}/tasks/{task.get('id', '')}",
            timestamp=task.get("updated"),
        )
        for task in data
    ]

File: connectors/unified-search/src/sources/vikunja.ts
import { SearchResultItem } from "../types";

const VIKUNJA_BASE_URL = process.env.VIKUNJA_BASE_URL ?? "https://vikunja.example.org";

/**
 * Queries the Vikunja `tasks/all?s=` API with the user's token relayed as-is
 * (study 2.2 line 391): only tasks from projects the user has access to
 * are returned, with no permission logic duplicated here.
 */
export async function searchVikunja(
  query: string,
  userToken: string,
  fetchImpl: typeof fetch = fetch
): Promise<SearchResultItem[]> {
  const response = await fetchImpl(
    `${VIKUNJA_BASE_URL}/api/v1/tasks/all?s=${encodeURIComponent(query)}`,
    {
      headers: { Authorization: `Bearer ${userToken}` },
    }
  );

  if (!response.ok) {
    throw new Error(`Vikunja search failed with status ${response.status}`);
  }

  const data = (await response.json()) as Array<{
    id?: number;
    title?: string;
    description?: string;
    updated?: string;
  }>;

  return (data ?? []).map((task) => ({
    source: "vikunja" as const,
    id: String(task.id ?? ""),
    title: task.title ?? "(task)",
    snippet: task.description,
    url: `${VIKUNJA_BASE_URL}/tasks/${task.id ?? ""}`,
    timestamp: task.updated,
  }));
}
